### DISCUSSION TITLE
Transaction.setIsolation()

### DISCUSSION THREAD

**Discussion by gavinking** [hibernate/hibernate-orm#5091] [Design Proposals]
@Sanne has pointed out that it's currently quite inconvenient to change the transaction isolation level, even though JDBC has an API for that.

I think it should be relatively straightforward to add a `setIsolation()` method to the `Transaction` interface.

WDYT?

**Comment by sebersole** [comment]
Actually from what I can tell, in JTA environments you generally cannot set isolation level per Connection as this is something managed at the DataSource level.  I'm not very keen on exposing this if we cannot support it consistently.  

One option is to simply set the isolation on the underlying Connection regardless of local or JTA transactions and let the user deal with the ramifications (inconsistent).

Another option would be to restrict this to just local transaction use cases.

Another consideration is how to handle this for databases which do not support certain isolations.

I guess the elephant in the room is what kind of guarantees we we envision here

**Reply by Sanne** [reply]
> Actually from what I can tell, in JTA environments you generally cannot set isolation level per Connection as this is something managed at the DataSource level. I'm not very keen on exposing this if we cannot support it consistently.

Right, the reason I highlighted this limitation is that it seems like a problem that would need to be solved across multiple specs. And I suspect it wouldn't be trivial to add such a feature to the Datasource implementations either, even if we only focused on the ones we care about.

But while it always puzzled *me* that such things aren't easy to do via idiomatic code in Java EE / Jakarta, I have rarely seen such requests from end users so I guess I'm the odd one - let's not do it just because I pointed this limitation out, I just wanted to remind about this limitation and see if you all have seen more user requests, as the lack of it puzzles me.

@scottmarlow has some more experience perhaps?

**Reply by gavinking** [reply]
I thought Stuart told us it's possible in Wildfly, even if not according to the spec.

**Reply by Sanne** [reply]
Right, I've certainly used it in the past on older JBoss AS versions. But I'm not sure how this would look in latest WildFly.

**Reply by beikov** [reply]
I can tell you, that back when I was a Hibernate user, I wanted to switch to SERIALIZABLE for certain transactions, so it definitely is something that might be interesting to some users. AFAIU, the only "standard" way to implement such a use case, is to create 2 data sources and 2 persistence units, which increases memory pressure.

**Reply by gavinking** [reply]
Yeah I mean I think it's clear that's a shit solution.

**Reply by scottmarlow** [reply]
https://vladmihalcea.com/a-beginners-guide-to-transaction-isolation-levels-in-enterprise-java/ was a good refresher for myself, mentioning it here for others that might like a refresher.

I have feedback on this topic and will share it soon.
